sdk/eventhubs/azure_messaging_eventhubs/src/common/recoverable_connection.rs

```diff
@@ -0,0 +1,1055 @@
+        self.recoverable_connection
+            .ensure_sender(&self.path)
+            .await?
+            .max_message_size()
+            .await
+    }
+}
+
+pub(crate) struct AmqpReceiverClient {
+    recoverable_connection: Arc<RecoverableConnection>,
+    source_url: Url,
+    message_source: AmqpSource,
+    receiver_options: AmqpReceiverOptions,
+    timeout: Option<Duration>,
+}
+
+#[cfg_attr(target_arch = "wasm32", async_trait::async_trait(?Send))]
+#[cfg_attr(not(target_arch = "wasm32"), async_trait::async_trait)]
+impl AmqpReceiverApis for AmqpReceiverClient {
+    async fn attach(
+        &self,
+        _session: &AmqpSession,
+        _source: impl Into<AmqpSource> + Send,
+        _options: Option<AmqpReceiverOptions>,
+    ) -> Result<()> {
+        unimplemented!("AmqpReceiverClient does not support attach operation");
+    }
+
+    async fn detach(self) -> azure_core::Result<()> {
+        unimplemented!("AmqpReceiverClient does not support detach operation");
+    }
+
+    async fn set_credit_mode(
+        &self,
+        _mode: azure_core_amqp::ReceiverCreditMode,
+    ) -> azure_core::Result<()> {
+        unimplemented!("AmqpReceiverClient does not support set_credit_mode operation");
+    }
+
+    async fn credit_mode(&self) -> azure_core::Result<azure_core_amqp::ReceiverCreditMode> {
+        unimplemented!("AmqpReceiverClient does not support credit_mode operation");
+    }
+
+    async fn receive_delivery(&self) -> azure_core::Result<azure_core_amqp::AmqpDelivery> {
+        let delivery = retry_azure_operation(
+            || async move {
+                let receiver = self
+                    .recoverable_connection
+                    .ensure_receiver(
+                        &self.source_url,
+                        &self.message_source,
+                        &self.receiver_options,
+                    )
+                    .await?;
+                if let Some(delivery_timeout) = self.timeout {
+                    select! {
+                        delivery = receiver.receive_delivery().fuse() => Ok(delivery),
+                        _ = azure_core::sleep::sleep(delivery_timeout).fuse() => {
+                             Err(azure_core::Error::new(
+                                AzureErrorKind::Io,
+                                Box::new(std::io::Error::from(std::io::ErrorKind::TimedOut))))
+                        },
+                    }?
+                } else {
+                    receiver.receive_delivery().await
+                }
+            },
+            &self.recoverable_connection.retry_options,
+            Some(RecoverableConnection::should_retry_receive_operation),
+        )
+        .await?;
+        Ok(delivery)
+    }
+
+    async fn accept_delivery(
+        &self,
+        _delivery: &azure_core_amqp::AmqpDelivery,
+    ) -> azure_core::Result<()> {
+        unimplemented!("AmqpReceiverClient does not support accept_delivery operation");
```

Comment:
With the number of unimplemented trait functions I see in this PR, sure maybe splitting traits up wouldn't be better? Generally, "if it compiles it works" and then pretty much flies in the face of that as these are all runtime panics. Almost like how serde splits up ser and de.